### ISSUE TITLE
Deploy Identityserver in a docker Linux container

### ISSUE THREAD

**Issue by OscarCanek** [aspnet/Home#2814]
### Issue / Steps to reproduce the problem

I'm trying to deploy Identityserver for high availability in a docker container with Linux but always have the following error when I try to login through mvc. I'm using a pfx certificate.

This is how I configure the DataProtection:

```
services.AddDataProtection()
                .PersistKeysToFileSystem(dirInfo)
                .ProtectKeysWithCertificate(new X509Certificate2(Configuration["server:pfxFilePath"], Configuration["server:pfxPassword"]))
                .SetApplicationName(!string.IsNullOrWhiteSpace(Configuration["appName"]) ? Configuration["appName"] : "XXXXXXXXX")
                .DisableAutomaticKeyGeneration();
```

### Relevant parts of the log file

```

An error occurred while trying to encrypt the provided data. Refer to the inner exception for more information.

InnerException: 
ClassName : "System.InvalidOperationException"
Message : "The key ring does not contain a valid default protection key. The data protection system cannot create a new key because auto-generation of keys is disabled."
Data : null
InnerException : null
HelpURL : null
StackTraceString : " at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingProvider.CreateCacheableKeyRingCore(DateTimeOffset now, IKey keyJustAdded) at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingProvider.GetCurrentKeyRingCore(DateTime utcNow) at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingBasedDataProtector.Protect(Byte[] plaintext)"
RemoteStackTraceString : null
RemoteStackIndex : 0
ExceptionMethod : null
HResult : -2146233079
Source : "Microsoft.AspNetCore.DataProtection"
WatsonBuckets : null
```
This is the complete message:

[Log message](https://github.com/IdentityServer/IdentityServer4/files/1640589/e.2.txt)

PD: use a json viewer to see it i.e. http://jsonviewer.stack.hu/

**Comment by blowdart** [comment]
Well the error message is pretty clear here, you disabled KeyGeneration and it wants to create a key because it doesn't have one, but it can't, because you disabled it.

So why did you disable key generation?

**Comment by blowdart** [comment]
This appears to have come from https://github.com/IdentityServer/IdentityServer4/issues/2005, so mentioned it in case others search on the same problem.

**Comment by OscarCanek** [comment]
I already have the keys at `dirInfo` and I don't need to create newones (I generate them with a separate console app using the same certificate I use here), so I have a key and when I run this aspnet app in windows no errors are shown. I think it's releated to the ProtectKeysWithCertificate, maybe in Linux don't support something because when I remove that it runs fine.

Some extra info:
- I generate the keys in a windows.
- I use a pfx certificate but also a PEM/CER certificate but nothing works

Is there a difference between this configuration running in Linux or windows?

**Comment by blowdart** [comment]
Well it appears it can't find your key ring, or decrypt it, so it's trying to generate new keys. I am presuming your keys haven't expired, and they're encrypted with the same certificate on Windows.

@morganbr / @bartonjs are there there any circumstances where encrypted XML decryption would work on Windows but no Linux.

**Comment by bartonjs** [comment]
@blowdart Not that I know of.  I don't see any tests for EncryptedXml which check to see what platform they're on.

**Comment by blowdart** [comment]
Ok @natemcmaster can you try to repo this please? Create a simple app that calls data protection once, with the keyring on disk and protected with an X509 cert and a fixed app name. Then also deploy it into a linux container, but have the container not generate keys and see what happens,

**Comment by blowdart** [comment]
@OscarCanek did you tweak the key expiry in any way? The default lifetime for keys is 14 days - is it possible the key you were using expired?

**Comment by OscarCanek** [comment]
I generated the keys based in the example here: [Key management](https://docs.microsoft.com/en-us/aspnet/core/security/data-protection/implementation/key-management)

key 1 (revoked)

```
<activationDate>2018-01-26T14:52:20.3073009Z</activationDate>
<expirationDate>2018-04-26T14:52:20.3073009Z</expirationDate>
```

The key above is generated by the following code:

```
// perform a protect operation to force the system to put at least
// one key in the key ring
services.GetDataProtector("Sample.KeyManager.v1").Protect("payload");
Console.WriteLine("Performed a protect operation.");
Thread.Sleep(2000);
```
And is revoked by this:

```
// list all keys in the key ring
var allKeys = keyManager.GetAllKeys();
// revoke all keys in the key ring
keyManager.RevokeAllKeys(DateTimeOffset.Now, reason: "Revocation reason here.");
Console.WriteLine("Revoked all existing keys.");
```

The following keys are generated after the code above:

key 2

```
<activationDate>2018-01-01T00:00:00-06:00</activationDate>
<expirationDate>2018-06-30T00:00:00-06:00</expirationDate>
```

key 3

```
<activationDate>2018-06-30T00:00:00-06:00</activationDate>
<expirationDate>2018-12-27T00:00:00-06:00</expirationDate>
```

key 4

```
<activationDate>2020-06-19T00:00:00-06:00</activationDate>
<expirationDate>2020-12-16T00:00:00-06:00</expirationDate>
```

key 5 

```
<activationDate>2019-06-25T00:00:00-06:00</activationDate>
<expirationDate>2019-12-22T00:00:00-06:00</expirationDate>
```

Key 6

```
<activationDate>2018-12-27T00:00:00-06:00</activationDate>
<expirationDate>2019-06-25T00:00:00-06:00</expirationDate>
```

key 6 

```
<activationDate>2019-12-22T00:00:00-06:00</activationDate>
<expirationDate>2020-06-19T00:00:00-06:00</expirationDate>
```

I think the `key 2` is being revoked by the RevokeAllKeys method because I use DateTimeOffset.Now and my start date of `key 2`  is older than `Now`. The interesting part is that in windows I haven't had problems. 

I will run some tests and I'll update this post.

**Comment by blowdart** [comment]
Ok so not that then :) Also tagging @GrabYourPitchforks

**Comment by natemcmaster** [comment]
I was able to reproduce the same error by generating keys with a cert, copying them into a container, and running the app. I believe this is a duplicate of #2321. The key ring is empty because the keys were protected by a certificate, but encrypted XML doesn't support decrypting with certs that are not in the X509 cert store.

To workaround this in ASP.NET Core 2.0.x, you can add the cert to your local X509Store inside your docker container like this:
```c#
var store = new X509Store(StoreName.My, StoreLocation.CurrentUser);
store.Open(OpenFlags.ReadWrite);
store.Add(cert);
store.Close();
``` 

You won't need this workaround in 2.1. We fixed this in https://github.com/aspnet/DataProtection/pull/299 by providing support for loading keys protected with a cert, even if that cert isn't in the X509Store. This will be released in ASP.NET Core 2.1.0-preview2. You can try an early preview of this by adding this to your csproj:
```xml
  <PropertyGroup>
    <RestoreAdditionalProjectSources>https://dotnet.myget.org/F/aspnetcore-dev/api/v3/index.json</RestoreAdditionalProjectSources>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.AspNetCore.DataProtection" Version="2.1.0-preview2-30143" />
  </ItemGroup>
```

@OscarCanek can you try one of these to confirm is solves your issue?

**Comment by natemcmaster** [comment]
Closing per previous comment. https://github.com/aspnet/Home/issues/2814#issuecomment-367145437. This issue is a duplicate of https://github.com/aspnet/Home/issues/2321 which will be resolved in 2.1.
